File: leetcode/medium/347_most_frequent_elements/most_frequent_elements.py
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        counts = {}
        for num in nums:
            if num in counts:
                counts[num] += 1
            else:
                counts[num] = 1
        counts = sorted(list(counts.items()), key=lambda x: -x[1])
        return list(map(lambda x: x[0], counts[:k]))

File: leetcode/medium/347_most_frequent_elements/most_frequent_elements.cpp
#include <algorithm>
#include <unordered_map>
#include <vector>

class Solution {
public:
  std::vector<int> topKFrequent(std::vector<int> &nums, int k) {
    std::unordered_map<int, int> counts;
    for (auto &num : nums) {
      auto maybe_num = counts.find(num);
      if (maybe_num == counts.end()) {
        counts.insert({num, 1});
      } else {
        maybe_num->second++;
      }
    }
    std::vector<std::pair<int, int>> count_pairs;
    for (auto &kv : counts) {
      count_pairs.push_back(kv);
    }
    std::sort(count_pairs.begin(), count_pairs.end(),
              [](std::pair<int, int> x, std::pair<int, int> y) {
                return x.second > y.second;
              });
    auto result_pairs =
        std::vector(count_pairs.begin(), count_pairs.begin() + k);
    std::vector<int> result;
    for (auto &pair : result_pairs) {
      result.push_back(pair.first);
    }
    return result;
  }
};
